File: DDD DSL/DDD DSL Design War FizzBuzzWhizz Reloaded/Code/Java/src/fayelab/ddd/fbwreloaded/complete/compiler/oo/Compiler.java
package fayelab.ddd.fbwreloaded.complete.compiler.oo;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

import fayelab.ddd.fbwreloaded.complete.compiler.oo.compiled.action.*;
import fayelab.ddd.fbwreloaded.complete.compiler.oo.compiled.predication.*;
import fayelab.ddd.fbwreloaded.complete.compiler.oo.compiled.rule.*;

public class Compiler
{
    private static Map<Predication.Type, Function<Integer, OOPredication>> predicationTypeAndFuncMap;
    
    static
    {
        predicationTypeAndFuncMap = new HashMap<>();
        predicationTypeAndFuncMap.put(Predication.Type.TIMES, Compiler::times);
        predicationTypeAndFuncMap.put(Predication.Type.CONTAINS, Compiler::contains);
        predicationTypeAndFuncMap.put(Predication.Type.ALWAYSTRUE, any -> alwaysTrue());
    }
    
    private static Map<Action.Type, Supplier<OOAction>> actionTypeAndFuncMap;
    
    static
    {
        actionTypeAndFuncMap = new HashMap<>();
        actionTypeAndFuncMap.put(Action.Type.TOFIZZ, Compiler::toFizz);
        actionTypeAndFuncMap.put(Action.Type.TOBUZZ, Compiler::toBuzz);
        actionTypeAndFuncMap.put(Action.Type.TOWHIZZ, Compiler::toWhizz);
        actionTypeAndFuncMap.put(Action.Type.TOSTR, Compiler::toStr);
    }
    
    private static Map<Rule.Type, Function<Rule, OORule>> ruleTypeAndFuncMap;
    
    static
    {
        ruleTypeAndFuncMap = new HashMap<>();
        ruleTypeAndFuncMap.put(Rule.Type.ATOM, rule -> compileAtom((Atom)rule.getData()));
        ruleTypeAndFuncMap.put(Rule.Type.OR, rule -> compileOr((Or)rule.getData()));
        ruleTypeAndFuncMap.put(Rule.Type.AND, rule -> compileAnd((And)rule.getData()));
    }
    
    public static OORule compile(Rule rule)
    {
        return ruleTypeAndFuncMap.get(rule.getType()).apply(rule);
    }
    
    static OOPredication compilePredication(Predication predication)
    {
        return predicationTypeAndFuncMap.get(predication.getType()).apply(predication.getParam());
    }
    
    static OOAction compileAction(Action action)
    {
        return actionTypeAndFuncMap.get(action.getType()).get();
    }
    
    static OOAtom compileAtom(Atom atom)
    {
        return new OOAtom(compilePredication(atom.getPredication()), compileAction(atom.getAction()));
    }

    static OOOr compileOr(Or or)
    {
        return new OOOr(compile(or.getRule1()), compile(or.getRule2()));
    }

    static OORule compileAnd(And and)
    {
        return new OOAnd(compile(and.getRule1()), compile(and.getRule2()));
    }
    
    private static OOPredication times(int base)
    {
        return new OOTimes(base);
    }

    private static OOPredication contains(int digit)
    {
        return new OOContains(digit);
    }

    private static OOPredication alwaysTrue()
    {
        return new OOAlwaysTrue();
    }
    
    private static OOAction toFizz()
    {
        return new OOToFizz();
    }
    
    private static OOAction toBuzz()
    {
        return new OOToBuzz();
    }
    
    private static OOAction toWhizz()
    {
        return new OOToWhizz();
    }
    
    private static OOAction toStr()
    {
        return new OOToStr();
    }
}

File: DDD DSL/DDD DSL Design War FizzBuzzWhizz Reloaded/Code/Python/complete/compiler/oo/compiler.py
from compiled.predication import *
from compiled.action import *
from compiled.rule import *

def c_times(base):
    return OOTimes(base)

def c_contains(digit):
    return OOContains(digit)

def c_alwaystrue(param):
    return OOAlwaysTrue()

PREDICATION_MAP = {'TIMES':c_times, 'CONTAINS':c_contains, 'ALWAYSTRUE':c_alwaystrue}

def compile_predication(predication):
    return PREDICATION_MAP[predication[0]](predication[1])

def c_tofizz():
    return OOToFizz()

def c_tobuzz():
    return OOToBuzz()

def c_towhizz():
    return OOToWhizz()

def c_tostr():
    return OOToStr()

ACTION_MAP = {'TOFIZZ':c_tofizz, 'TOBUZZ':c_tobuzz, 'TOWHIZZ':c_towhizz, 'TOSTR':c_tostr}

def compile_action(action):
    return ACTION_MAP[action]()

def compile_atom(atom_rule):
    return OOAtom(compile_predication(atom_rule[1]), compile_action(atom_rule[2]))

def compile_or(or_rule):
    return OOOR(compile(or_rule[1]), compile(or_rule[2]))

def compile_and(and_rule):
    return OOAND(compile(and_rule[1]), compile(and_rule[2]))

RULE_MAP = {'ATOM':compile_atom, 'OR':compile_or, 'AND':compile_and}

def compile(rule):
    return RULE_MAP[rule[0]](rule)